proknow/Audit.py

```diff
@@ -0,0 +1,159 @@
+            A :class:`proknow.Audit.AuditResultsPage` object, representing a page of query results
+
+        Raises:
+            :class:`proknow.Exceptions.HttpError`: If the HTTP request generated an error.
+        """
+
+        self.options = {}
+        if page_size is not None: self.options["page_size"] = page_size
+        if "start_time" in kwargs : self.options["start_time"] = kwargs["start_time"].isoformat()
+        if "end_time" in kwargs: self.options["end_time"] = kwargs["end_time"].isoformat()
+        if "user_id" in kwargs: self.options["user_id"] = kwargs["user_id"] 
+        if "user_name" in kwargs: self.options["user_name"] = kwargs["user_name"]
+        if "patient_id" in kwargs: self.options["patient_id"] = kwargs["patient_id"]
+        if "patient_name" in kwargs: self.options["patient_name"] = kwargs["patient_name"]
+        if "patient_mrn" in kwargs: self.options["patient_mrn"] = kwargs["patient_mrn"]
+        if "workspace_id" in kwargs: self.options["workspace_id"] = kwargs["workspace_id"]
+        if "workspace_name" in kwargs: self.options["workspace_name"] = kwargs["workspace_name"]
+        if "collection_id" in kwargs: self.options["collection_id"] = kwargs["collection_id"]
+        if "resource_id" in kwargs: self.options["resource_id"] = kwargs["resource_id"]
+        if "resource_name" in kwargs: self.options["resource_name"] = kwargs["resource_name"] 
+        if "classification" in kwargs: self.options["classification"] = kwargs["classification"].upper() 
+        if "uri" in kwargs: self.options["uri"] = kwargs["uri"] 
+        if "user_agent" in kwargs: self.options["user_agent"] = kwargs["user_agent"]
+        if "ip_address" in kwargs: self.options["ip_address"] = kwargs["ip_address"]
+
+        if "types" in kwargs: 
+            if not isinstance(kwargs["types"], list):
+                self.options["types"] = [kwargs["types"]]
+            else:
+                self.options["types"] = kwargs["types"]
+
+        if "methods" in kwargs: 
+            if not isinstance(kwargs["methods"], list):
+                self.options["methods"] = [kwargs["methods"].upper()]
+            else:
+                self.options["methods"] = [x.upper() for x in kwargs["methods"]]
+
+        if "status_codes" in kwargs: 
+            if not isinstance(kwargs["status_codes"], list):
+                self.options["status_codes"] = [kwargs["status_codes"]]
+            else:
+                self.options["status_codes"] = kwargs["status_codes"]
+
+        if "text" in kwargs: 
+            if not isinstance(kwargs["text"], list):
+                self.options["text"] = [kwargs["text"]]
+            else:
+                self.options["text"] = kwargs["text"]
+
+        self.auditResultsPage = self._query()
+
+        self.options['page_number'] = 0
+        if self.auditResultsPage.total > 0:
+            self.options['first_id'] = self.auditResultsPage.items[0]['id']
+
+        return self.auditResultsPage
```

Comment:
I don't think the Audit service should be maintaining the options. Think about this case:

```python
clinicalResults = pk.audit.query(workspace_name="Clinical")
researchResults = pk.audit.query(workspace_name="Research")
nextClinicalResults = clinicalResults.next()
nextResearchResults = researchResults.next()
```

Because the options are being stored in the Audit instance, `self.options` for the Clinical workspace will be overwritten by the `self.options` for the Research workspace, leading to some strange results. The options should be constructed with every `query` call, passed to `_query`, and then passed to the `AuditResultsPage` constructor.